Page transitions replaced by ZoomPageTransitionsBuilder (#6646)
flutter/website: src/release/breaking-changes/page-transition-replaced-by-ZoomPageTransitionBuilder.md

---
title: Page transitions replaced by ZoomPageTransitionsBuilder
description: Using the latest page transition instead of the old one.
---

## Summary

> In order to keep libraries follow the latest OEM behavior,
> we've replaced the default page transition builders from
> `FadeUpwardsPageTransitionsBuilder` to
> `ZoomPageTransitionsBuilder` on all platforms excluding iOS and macOS.

Reviewer: ```suggestion
In order to ensure that libraries follow the latest OEM behavior,
the default page transition builders now use
`ZoomPageTransitionsBuilder` on all platforms (excluding iOS and macOS)
instead of `FadeUpwardsPageTransitionsBuilder`.
```

## Context

> `FadeUpwardsPageTransitionsBuilder` came with the first
> Flutter release, which defines a page transition that's
> similar to the one provided by Android O.

Reviewer: ```suggestion
The `FadeUpwardsPageTransitionsBuilder` (provided with the first
Flutter release), defined a page transition that's
similar to the one provided by Android O. This page transitions builder
will eventually be deprecated on Android, as per Flutter's
[deprecation policy]({{site.url}}/resources/compatibility#deprecation-policy).
```

> `ZoomPageTransitionsBuilder` defines a page transition that's
> similar to the one provided by Android Q and R.

Reviewer: ```suggestion
`ZoomPageTransitionsBuilder`, the new page transition builder for Android, Linux, and Windows,
defines a page transition that's similar to the one provided by Android Q and R.
```

According to the [Style guide for Flutter repo][],
the framework will follow the latest OEM behavior.
Page transition builders using `FadeUpwardsPageTransitionsBuilder`
are all switched to the `ZoomPageTransitionsBuilder`.
When the current `TargetPlatform` doesn't have
`PageTransitionsBuilder` defined in the `ThemeData.pageTransitionsTheme`,

> `ZoomPageTransitionsBuilder` will be use as default.

Reviewer: ```suggestion
`ZoomPageTransitionsBuilder` is used as the default.
```

`FadeUpwardsPageTransitionsBuilder` to
`ZoomPageTransitionsBuilder` for `TargetPlatform.android`,
`TargetPlatform.linux` and `TargetPlatform.windows`.

## Migration guide

If you want to switch back to the previous page transition builder
(`FadeUpwardsPageTransitionsBuilder`), you should define builders
explicitly for the target platforms.

Code before migration:

<!-- skip -->
```dart
MaterialApp(
  theme: ThemeData(primarySwatch: Colors.blue),
)
```

Code after migration:

<!-- skip -->
```dart
MaterialApp(
  theme: ThemeData(
    pageTransitionsTheme: const PageTransitionsTheme(
      builders: <TargetPlatform, PageTransitionsBuilder>{
        TargetPlatform.android: FadeUpwardsPageTransitionsBuilder(), // Apply this to every platforms you need.
      },
    ),
  ),
)
```

If you want to apply the same page transition builder to all platforms:

<!-- skip -->
```dart
MaterialApp(
  theme: ThemeData(
    pageTransitionsTheme: PageTransitionsTheme(
      builders: Map<TargetPlatform, PageTransitionsBuilder>.fromIterable(
        TargetPlatform.values,
        value: (dynamic _) => const FadeUpwardsPageTransitionsBuilder(),
      ),
    ),
  ),
)

```

### Tests migration

> If you used to find widgets but failed with *Too many elements*
> using the new transition, and errors are like these:

Reviewer: ```suggestion
If you used to try to find widgets but failed with *Too many elements*
using the new transition, and saw errors similar to the following:
```

```
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following StateError was thrown running a test:
Bad state: Too many elements

When the exception was thrown, this was the stack:
#0      Iterable.single (dart:core/iterable.dart:656:24)
#1      WidgetController.widget (package:flutter_test/src/controller.dart:69:30)
#2      main.<anonymous closure> (file:///path/to/your/test.dart:1:2)
```

You should migrate your tests by using the
`descendant` scope for `Finder`s with the specific widget type.
Below is the example of `DataTable`'s test:

> Test before migrate:

Reviewer: ```suggestion
Test before migration:
```

```dart
final Finder finder = find.widgetWithIcon(Transform, Icons.arrow_upward);
```

> Test after migrate:

Reviewer: ```suggestion
Test after migration:
```

```dart
final Finder finder = find.descendant(
  of: find.byType(DataTable),
  matching: find.widgetWithIcon(Transform, Icons.arrow_upward),
);
```

> Widgets that typically needs to migrate the finder scope are:

Reviewer: ```suggestion
Widgets that typically need to migrate the finder scope are:
```

`Transform`, `FadeTransition`, `ScaleTransition`, and `ColoredBox`.

## Timeline

Landed in version: v2.9.0-1.0.pre<br>

> In stable release: Not yet

Reviewer 2: I'm guessing this documentation should be held until this is in stable.

Author: The PR should be held until the change is rolled. But shouldn't a breaking change note released along with the change? The *In stable release* should update after.

Reviewer 2: Ahh, yes you are right.

Reviewer: Yes, @AlexV525 is correct.  :)

  ),
)

```

### Tests migration

If you used to try to find widgets but failed with *Too many elements*
using the new transition, and saw errors similar to the following:

```
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following StateError was thrown running a test:
Bad state: Too many elements

When the exception was thrown, this was the stack:
#0      Iterable.single (dart:core/iterable.dart:656:24)
#1      WidgetController.widget (package:flutter_test/src/controller.dart:69:30)
#2      main.<anonymous closure> (file:///path/to/your/test.dart:1:2)
```

You should migrate your tests by using the
`descendant` scope for `Finder`s with the specific widget type.
Below is the example of `DataTable`'s test:

Test before migration:
```dart
final Finder finder = find.widgetWithIcon(Transform, Icons.arrow_upward);
```

Test after migration:
```dart
final Finder finder = find.descendant(
  of: find.byType(DataTable),
  matching: find.widgetWithIcon(Transform, Icons.arrow_upward),
);
```

Widgets that typically need to migrate the finder scope are:
`Transform`, `FadeTransition`, `ScaleTransition`, and `ColoredBox`.

## Timeline

> Landed in version: v2.9.0-1.0.pre<br>

Author: ```suggestion
Landed in version: v2.13.0-1.0.pre<br>
```

In stable release: Not yet

## References

API documentation:
* [`ZoomPageTransitionsBuilder`][]
* [`FadeUpwardsPageTransitionsBuilder`][]
* [`PageTransitionsTheme`][]

Relevant issues:
* [Issue 43277][]

Relevant PR:

> * [PR 96375][]

Author: ```suggestion
* [PR 100812][]
```

[`ZoomPageTransitionsBuilder`]: {{site.api}}/flutter/material/ZoomPageTransitionsBuilder-class.html
[`FadeUpwardsPageTransitionsBuilder`]: {{site.api}}/flutter/material/FadeUpwardsPageTransitionsBuilder-class.html
[`PageTransitionsTheme`]: {{site.api}}/flutter/material/PageTransitionsTheme-class.html
[Issue 43277]: {{site.repo.flutter}}/issues/43277

> [PR 96375]: {{site.repo.flutter}}/pull/96375

Author: ```suggestion
[PR 100812]: {{site.repo.flutter}}/pull/100812
```

---
title: Page transitions replaced by ZoomPageTransitionsBuilder
description: Using the latest page transition instead of the old one.
---

## Summary

In order to keep libraries follow the latest OEM behavior,
we've replaced the default page transition builders from
`FadeUpwardsPageTransitionsBuilder` to

> `ZoomPageTransitionsBuilder` on all platforms excluded iOS and macOS.

Reviewer 3: "excluding"

## Context

`FadeUpwardsPageTransitionsBuilder` came with the first
Flutter release, which defines a page transition that's
similar to the one provided by Android O.

`ZoomPageTransitionsBuilder` defines a page transition that's
similar to the one provided by Android Q and R.

According to the [Style guide for Flutter repo][], we've switched

> the default page transitions which mainly for all platforms excluded iOS and macOS.

Reviewer 3: not sure what this sentence means

Author: It's the same as L20-23, I think it might be bad to use the same sentence.

Reviewer 3: I don't think lines 22 and 23 here are grammatically correct English. I would suggest alternative wording but I'm not sure what they're trying to say so I'm not sure how to fix it.

> When the `PageTransitionsBuilder` is not found on the current platform,

Reviewer 3: what does "is not found" mean?

Author: That is https://github.com/flutter/flutter/pull/96375/commits/d5c96054ebb763b9374a74ec7ebc5840036645a6/packages/flutter/lib/src/material/page_transitions_theme.dart#L598-599

```dart
final PageTransitionsBuilder matchingBuilder =
    builders[platform] ?? const ZoomPageTransitionsBuilder();
```

Reviewer 3: sorry to clarify when i ask a question like that what i really mean is "please update the text to answer the following question: ..." :-)

> `ZoomPageTransitionsBuilder` will be the default as well.

Reviewer 3: strike "as well"

[Style guide for Flutter repo]: {{site.repo.flutter}}/wiki/Style-guide-for-Flutter-repo

## Description of change

> `PageTransitionsBuilder`s that defined in

Reviewer 3: strike "that"

`PageTransitionsTheme._defaultBuilders` have changed from
`FadeUpwardsPageTransitionsBuilder` to
`ZoomPageTransitionsBuilder` for `TargetPlatform.android`,
`TargetPlatform.linux` and `TargetPlatform.windows`.

## Migration guide

If you want to switch back to the previous page transition builder
(`FadeUpwardsPageTransitionsBuilder`), you should define builders
explicitly for the target platforms.

Code before migration:

<!-- skip -->
```dart
MaterialApp(
  theme: ThemeData(primarySwatch: Colors.blue),
)
```

Code after migration:

<!-- skip -->
```dart
MaterialApp(
  theme: ThemeData(
    primarySwatch: Colors.blue,
    pageTransitionsTheme: const PageTransitionsTheme(
      builders: <TargetPlatform, PageTransitionsBuilder>{

>         TargetPlatform.android: FadeUpwardsPageTransitionsBuilder(), // Apply this to every platforms you need.

Reviewer 3: is there a way to apply it to all platforms at once rather than having to list them all?

Author: Should we indicate that here?
